user/app_main.c

```diff
@@ -2,14 +2,200 @@
 #include <pthread.h>
 
 #include <driver.h>
+#include <memzone.h>
+#include <pthread.h>
+
+#include <nfp_rtsym.h>
+#include <nfp_cpp.h>
+#include <nfp_net_pmd.h>
+#include <rte_byteorder.h>
+
+#define RX_HEAD_SYM "i32._rx_head"
+#define RX_TAIL_SYM "i32._rx_tail"
+#define TX_HEAD_SYM "i32._tx_head"
+#define TX_TAIL_SYM "i32._tx_tail"
+#define RX_BUF_START_SYM "i32._rx_buf_start"
+#define RX_BUF_LEN_SYM "i32._rx_buf_len"
+#define TX_BUF_START_SYM "i32._tx_buf_start"
+#define TX_BUF_LEN_SYM "i32._tx_buf_len"
+
+#define START_SYM "i32._start"
+
+#define UDP_PACKET_SIZE 64
+#define BUF_SIZE 512
+
+#define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))
 
 extern int nfp_cpp_dev_main(struct nfp_cpp* cpp);
+const struct memzone* mzone_rx;
+const struct memzone* mzone_tx;
+
+uint8_t *rx_head_sym, *rx_tail_sym, *tx_head_sym, *tx_tail_sym, *rx_buf_start_sym, *rx_buf_len_sym, *tx_buf_start_sym, *tx_buf_len_sym;
+uint8_t *start_sym;
+
+uint64_t rx_head_virt;
+uint64_t tx_tail_virt;
+uint64_t rx_buf_start, tx_buf_start, rx_buf_len, tx_buf_len;
+
+static inline uint64_t RD(volatile void *addr)
+{
+	const volatile uint32_t *p = addr;
+	uint32_t low, high;
+
+	low = nn_readl((volatile const void *)(p + 1));
```

Comment:
Pointer arithmetic on `void*` is ill-defined.
May be this is the right thing to do?
`low = nn_readl(((volatile uin32_t*)p) + 1)`

Reply:
My bad! p is `uint32_t*`